node/src/Pos/Block/Logic/Header.hs

```diff
@@ -328,35 +327,96 @@ getHeadersOlderExp upto = do
+    unlessM (isJust <$> (DB.blkGetHeader @ssc newer)) $
+        throwError "getHeadersRange: can't find newer-older header"
+    whenJust depthLimitM $ \depthLimit ->
+        when (depthLimit < 1) $
+        throwError $
+        sformat ("getHeadersRange: depthLimit is "%int%
+                 ", we can't return the single requested header "%build)
+                depthLimit
+                newer
+    pure $ OldestFirst $ one newer
+getHeadersRange depthLimitM older newer = runExceptT $ do
     -- oldest and newest blocks do exist
-    start <- MaybeT $ DB.blkGetHeader @ssc newer
-    end   <- MaybeT $ DB.blkGetHeader @ssc older
-    guard $ getEpochOrSlot start >= getEpochOrSlot end
-    let lowerBound = getEpochOrSlot end
-    if newer == older
-    then pure $ one newer
-    else loadHeadersDo lowerBound (one newer) $ start ^. prevBlockL
+    newerHd <- fromMaybeM "can't retrieve newer header" $ DB.blkGetHeader @ssc newer
+    olderHd <- fromMaybeM "can't retrieve older header" $ DB.blkGetHeader @ssc older
+    let olderD = olderHd ^. difficultyL
+    let newerD = newerHd ^. difficultyL
+
+    -- Proving newerD >= olderD
+    let newerOlderF = "newer: "%build%", older: "%build
+    when (newerD == olderD) $
+        throwError $
+        sformat ("getHeadersRange: newer and older headers have "%
+                 "the same difficulty, but are not equal. "%newerOlderF)
+                newerHd olderHd
+    when (newerD < olderD) $
+        throwError $
+        sformat ("getHeadersRange: newer header is less dificult than older one. "%
+                 newerOlderF)
+                newerHd olderHd
+
+    -- How many epochs does this range cross.
+    let genDiff :: Int
+        genDiff = fromIntegral $ newerHd ^. epochIndexL - olderHd ^. epochIndexL
+    -- Number of blocks is difficulty difference + number of genesis blocks.
+    -- depthDiff + 1 is length of a list we'll return.
+    let depthDiff :: Word
+        depthDiff = fromIntegral $ genDiff + fromIntegral (newerD - olderD)
+
+    whenJust depthLimitM $ \depthLimit ->
+        when (depthDiff + 1 > depthLimit) $
+        throwError $
+        sformat ("getHeadersRange: requested "%int%" headers, but depthLimit is "%
+                 int%". Headers: "%newerOlderF)
+                depthDiff
+                depthLimit
+                newerHd olderHd
+
+    -- We load these depthDiff blocks.
+    let cond curHash _depth = curHash /= newer
+
+    -- This is [oldest..newest) headers, oldest first
+    allExceptNewest <- GS.loadHashesUpWhile older cond
+
+    -- Sometimes we will get an empty list, if we've just switched the
+    -- branch (after first checks are performed here) and olderHd is
+    -- no longer in the main chain.
+    -- CSL-1950 We should use snapshots here.
+    when (null $ allExceptNewest ^. _Wrapped) $ throwError $
+        "getHeadersRange: loaded 0 headers though checks passed. " <>
+        "May be (very rare) concurrency problem, just retry"
```

Comment:
Ah, this is unfortunate. I take it this bug was always there?

I recommend putting a short explanation in the log message... maybe "branch may have forked in the midst of loading blocks" or something similar.

Reply:
There's an issue for it: https://iohk.myjetbrains.com/youtrack/issue/CSL-1950

Reply:
Obligatory https://en.wikipedia.org/wiki/Murphy%27s_law

Reply:
@avieth sure, i _obviously_ thought of it when writing the comment, sanity check passed. What do we do if error kicks in:
1. Locating the error message in codebase (`grep`, takes 15 seconds in total)
2. Finding comment about YT issue (3 seconds)
3. Deciding if the fix is needed and pushing it (1h at most)

:)

Reply:
> Deciding if the fix is needed and pushing it (1h at most)

Just 1 hour to implement a fix for this and push it? I highly doubt that.

Reply:
It's a symbolic number, the main point is that we don't have any evidence that this issue may be problematic. There are other different places (one or two, i believe) in CSL that also have this assumption.

Reply:
Sure, I'm not saying it has to be fixed here, since it's not a new bug.